src/orchestrator/streaming/aggregator.rs

```diff
@@ -0,0 +1,484 @@
+            start_index: Some(chunk.start_index as u32),
+            processed_index: Some(chunk.processed_index as u32),
+            tokens: Some(tokens),
+            // Populate fields from last response or default
+            ..generations.last().cloned().unwrap_or_default()
+        };
+        result.token_classification_results.output = Some(detections);
+        if input_start_index == 0 {
+            // Get input_token_count and seed from first generation message
+            let first = generations.first().unwrap();
+            result.input_token_count = first.input_token_count;
+            result.seed = first.seed;
+            // Get input_tokens from second generation message (if specified)
+            let input_tokens = if let Some(second) = generations.get(1) {
+                second.input_tokens.clone()
+            } else {
+                Some(Vec::default())
+            };
+            result.input_tokens = input_tokens;
+        }
+
+        // Send result to result channel
+        let _ = self.result_tx.send(Ok(result)).await;
+    }
+}
+
+/// [`ResultActor`] handle.
+struct ResultActorHandle {
+    tx: mpsc::Sender<ResultActorMessage>,
+}
+
+impl ResultActorHandle {
+    pub fn new(
+        generation_actor: Arc<GenerationActorHandle>,
+        result_tx: mpsc::Sender<Result<ClassifiedGeneratedTextStreamResult, Error>>,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel(32);
+        let mut actor = ResultActor::new(rx, generation_actor, result_tx);
+        tokio::spawn(async move { actor.run().await });
+        Self { tx }
+    }
+
+    pub async fn send(&self, chunk: Chunk, detections: Detections) {
+        let msg = ResultActorMessage((chunk, detections));
+        let _ = self.tx.send(msg).await;
+    }
+}
+
+#[derive(Debug)]
+struct AggregationActorMessage {
+    pub chunk: Chunk,
+    pub detections: Detections,
+}
+
+/// Aggregates detections and sends them to [`ResultActor`].
+struct AggregationActor {
+    rx: mpsc::Receiver<AggregationActorMessage>,
+    result_actor: ResultActorHandle,
+    tracker: Tracker,
+    n_detectors: usize,
+}
+
+impl AggregationActor {
+    pub fn new(
+        rx: mpsc::Receiver<AggregationActorMessage>,
+        result_actor: ResultActorHandle,
+        n_detectors: usize,
+    ) -> Self {
+        let tracker = Tracker::new();
+        Self {
+            rx,
+            result_actor,
+            tracker,
+            n_detectors,
+        }
+    }
+
+    async fn run(&mut self) {
+        while let Some(msg) = self.rx.recv().await {
+            self.handle(msg).await;
+        }
+    }
+
+    async fn handle(&mut self, msg: AggregationActorMessage) {
+        // TODO: support overlapping spans from different chunkers?
+        let chunk = msg.chunk;
+        let detections = msg.detections;
+
+        // Add to tracker
+        let span: Span = (chunk.start_index, chunk.processed_index);
+        self.tracker.push((span, detections));
```

Comment:
This implementation should provide the same results as the original one, and our testing thus far has validated that. It just takes a different approach for how we get there.

> So we end up receiving response for the 2nd chunk from both of the detectors before we get "complete" response for 1st chunk. 

> So I think in current implementation, we would end up sending the response to our user for the 2nd chunk since its complete already right ? This was sort of being tracked with the processed_index logic in previous implementation.

The `AggregationActor` is responsible for the bookkeeping and it will not send any responses (via the `ResultActor`) until responses from all detectors for a given chunk have been captured. The logic [here](https://github.com/declark1/fms-guardrails-orchestrator/blob/aggregator-refactor/src/orchestrator/streaming/aggregator.rs#L226-L233) checks this.

Are you suggesting we need to add in additional logic to handle an out-of-order scenario? I will try to think of a way to test this...

Reply:
Yep, exactly, am suggesting we need additional logic around this